File: packages/core/src/routes/resumes.ts
/**
 * Resume routes — thin HTTP layer over ResumeService.
 */

import { Hono } from 'hono'
import type { Database } from 'bun:sqlite'
import type { Services } from '../services'
import { mapStatusCode } from './server'
import type { CreateResume, UpdateResume, AddResumeEntry } from '../types'

export function resumeRoutes(services: Services, db: Database) {
  const app = new Hono()

  app.post('/resumes', async (c) => {
    const body = await c.req.json<CreateResume & { template_id?: string }>()

    if (body.template_id) {
      const result = services.templates.createResumeFromTemplate({
        ...body,
        template_id: body.template_id,
      })
      if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
      return c.json({ data: result.data }, 201)
    }

    const result = services.resumes.createResume(body)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data }, 201)
  })

  app.get('/resumes', (c) => {
    const offset = Math.max(0, parseInt(c.req.query('offset') ?? '0', 10) || 0)
    const limit = Math.min(200, Math.max(1, parseInt(c.req.query('limit') ?? '50', 10) || 50))

    const result = services.resumes.listResumes(offset, limit)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data, pagination: result.pagination })
  })

  app.get('/resumes/:id', (c) => {
    const result = services.resumes.getResume(c.req.param('id'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  app.patch('/resumes/:id', async (c) => {
    const body = await c.req.json<UpdateResume>()
    const result = services.resumes.updateResume(c.req.param('id'), body)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  app.delete('/resumes/:id', (c) => {
    const result = services.resumes.deleteResume(c.req.param('id'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.body(null, 204)
  })

  app.post('/resumes/:id/entries', async (c) => {
    const body = await c.req.json<AddResumeEntry>()
    const result = services.resumes.addEntry(c.req.param('id'), body)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data }, 201)
  })

  // Reorder must come before :entryId to avoid "reorder" matching as a param
  app.patch('/resumes/:id/entries/reorder', async (c) => {
    const body = await c.req.json<{ entries: Array<{ id: string; section_id: string; position: number }> }>()
    const result = services.resumes.reorderEntries(c.req.param('id'), body.entries)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: null })
  })

  app.patch('/resumes/:id/entries/:entryId', async (c) => {
    const body = await c.req.json<{ content?: string | null; section_id?: string; position?: number; notes?: string | null }>()
    const result = services.resumes.updateEntry(c.req.param('id'), c.req.param('entryId'), body)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  app.delete('/resumes/:id/entries/:entryId', (c) => {
    const result = services.resumes.removeEntry(c.req.param('id'), c.req.param('entryId'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.body(null, 204)
  })

  // ── Section CRUD ────────────────────────────────────────────────────

  app.post('/resumes/:id/sections', async (c) => {
    const body = await c.req.json<{ title: string; entry_type: string; position?: number }>()
    const result = services.resumes.createSection(c.req.param('id'), body)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data }, 201)
  })

  app.get('/resumes/:id/sections', (c) => {
    const result = services.resumes.listSections(c.req.param('id'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  app.patch('/resumes/:id/sections/:sectionId', async (c) => {
    const body = await c.req.json<{ title?: string; position?: number }>()
    const result = services.resumes.updateSection(c.req.param('id'), c.req.param('sectionId'), body)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  app.delete('/resumes/:id/sections/:sectionId', (c) => {
    const result = services.resumes.deleteSection(c.req.param('id'), c.req.param('sectionId'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.body(null, 204)
  })

  // ── Resume Skills (within a section) ───────────────────────────────

  app.post('/resumes/:id/sections/:sectionId/skills', async (c) => {
    const body = await c.req.json<{ skill_id: string }>()
    const result = services.resumes.addSkill(c.req.param('id'), c.req.param('sectionId'), body.skill_id)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data }, 201)
  })

  app.get('/resumes/:id/sections/:sectionId/skills', (c) => {
    const result = services.resumes.listSkillsForSection(c.req.param('id'), c.req.param('sectionId'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  app.delete('/resumes/:id/sections/:sectionId/skills/:skillId', (c) => {
    const result = services.resumes.removeSkill(c.req.param('id'), c.req.param('sectionId'), c.req.param('skillId'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.body(null, 204)
  })

  app.patch('/resumes/:id/sections/:sectionId/skills/reorder', async (c) => {
    const body = await c.req.json<{ skills: Array<{ skill_id: string; position: number }> }>()
    const result = services.resumes.reorderSkills(c.req.param('id'), c.req.param('sectionId'), body.skills)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: null })
  })

  // -- Save as Template ---------------------------------------------------

  app.post('/resumes/:id/save-as-template', async (c) => {
    const body = await c.req.json<{ name: string; description?: string }>()
    const result = services.templates.saveAsTemplate(
      c.req.param('id'),
      body.name,
      body.description,
    )
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data }, 201)
  })

  app.get('/resumes/:id/gaps', (c) => {
    const result = services.resumes.analyzeGaps(c.req.param('id'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  // ── IR endpoint ──────────────────────────────────────────────────

  app.get('/resumes/:id/ir', (c) => {
    const result = services.resumes.getIR(c.req.param('id'))
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  // ── Header & Override endpoints ────────────────────────────────────

  app.patch('/resumes/:id/header', async (c) => {
    const body = await c.req.json()
    const result = services.resumes.updateHeader(c.req.param('id'), body)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  app.patch('/resumes/:id/markdown-override', async (c) => {
    const body = await c.req.json<{ content: string | null }>()
    const result = services.resumes.updateMarkdownOverride(c.req.param('id'), body.content)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  app.patch('/resumes/:id/latex-override', async (c) => {
    const body = await c.req.json<{ content: string | null }>()
    const result = services.resumes.updateLatexOverride(c.req.param('id'), body.content)
    if (!result.ok) return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  // ── Tagline endpoints (Phase 92) ──────────────────────────────────

  /**
   * Return the resume's current tagline state: the generated tagline
   * (from linked JDs), the user's override (if any), the resolved value
   * (override takes precedence), and a flag indicating whether an override
   * is present.
   */
  app.get('/resumes/:id/tagline', (c) => {
    const row = db
      .query(
        'SELECT generated_tagline, tagline_override FROM resumes WHERE id = ?',
      )
      .get(c.req.param('id')) as {
        generated_tagline: string | null
        tagline_override: string | null
      } | null

    if (!row) {
      return c.json(
        { error: { code: 'NOT_FOUND', message: 'Resume not found' } },
        404,
      )
    }

    const has_override = !!(row.tagline_override && row.tagline_override.trim().length > 0)
    const resolved = row.tagline_override ?? row.generated_tagline ?? ''

    return c.json({
      data: {
        generated_tagline: row.generated_tagline,
        tagline_override: row.tagline_override,
        resolved,
        has_override,
      },
    })
  })

  /** Force a tagline regeneration from currently linked JDs. */
  app.post('/resumes/:id/tagline/regenerate', async (c) => {
    const { regenerateResumeTagline } = await import('../services/tagline-service')
    const result = regenerateResumeTagline(db, c.req.param('id'))
    if (!result) {
      return c.json(
        { error: { code: 'NOT_FOUND', message: 'Resume not found' } },
        404,
      )
    }
    return c.json({ data: result })
  })

  /**
   * Set or clear the tagline override. Pass `{ content: string }` to set,
   * `{ content: null }` to clear the override (falls back to generated).
   */
  app.patch('/resumes/:id/tagline-override', async (c) => {
    const body = await c.req.json<{ content: string | null }>()
    const resumeId = c.req.param('id')

    const exists = db
      .query('SELECT id FROM resumes WHERE id = ?')
      .get(resumeId) as { id: string } | null
    if (!exists) {
      return c.json(
        { error: { code: 'NOT_FOUND', message: 'Resume not found' } },
        404,
      )
    }

    const normalized = body.content && body.content.trim().length > 0 ? body.content : null
    db.run(
      `UPDATE resumes
       SET tagline_override = ?,
           updated_at = strftime('%Y-%m-%dT%H:%M:%SZ', 'now')
       WHERE id = ?`,
      [normalized, resumeId],
    )

    const row = db
      .query('SELECT generated_tagline, tagline_override FROM resumes WHERE id = ?')
      .get(resumeId) as { generated_tagline: string | null; tagline_override: string | null }

    const has_override = !!(row.tagline_override && row.tagline_override.trim().length > 0)
    const resolved = row.tagline_override ?? row.generated_tagline ?? ''

    return c.json({
      data: {
        generated_tagline: row.generated_tagline,
        tagline_override: row.tagline_override,
        resolved,
        has_override,
      },
    })
  })

  // ── PDF endpoint ──────────────────────────────────────────────────

  app.post('/resumes/:id/pdf', async (c) => {
    let latex: string | undefined
    try {
      const body = await c.req.json()
      latex = body.latex
    } catch {
      // No body is fine -- compile from IR
    }

    const result = await services.resumes.generatePDF(c.req.param('id'), latex)
    if (!result.ok) {
      const code = result.error.code
      const status = code === 'TECTONIC_NOT_AVAILABLE' ? 501
                   : code === 'TECTONIC_TIMEOUT' ? 504
                   : code === 'LATEX_COMPILE_ERROR' ? 422
                   : mapStatusCode(code)
      return c.json({ error: result.error }, status as any)
    }

    return new Response(result.data, {
      status: 200,
      headers: {
        'Content-Type': 'application/pdf',
        'Content-Disposition': 'inline; filename="resume.pdf"',
      },
    })
  })

  // ── Linked JDs (reverse lookup) ──────────────────────────────────

  app.get('/resumes/:id/job-descriptions', (c) => {
    const resumeId = c.req.param('id')

    // Verify resume exists
    const resume = services.resumes.getResume(resumeId)
    if (!resume.ok) return c.json({ error: resume.error }, mapStatusCode(resume.error.code))

    const rows = db
      .query(`
        SELECT jd.id AS job_description_id, jd.title, o.name AS organization_name,
               jd.status, jd.location, jd.salary_range,
               jdr.created_at, jd.created_at AS jd_created_at
        FROM job_description_resumes jdr
        JOIN job_descriptions jd ON jd.id = jdr.job_description_id
        LEFT JOIN organizations o ON o.id = jd.organization_id
        WHERE jdr.resume_id = ?
        ORDER BY jdr.created_at DESC
      `)
      .all(resumeId)

    return c.json({ data: rows })
  })

  // ── Contact reverse lookup ──────────────────────────────────────────
  app.get('/resumes/:id/contacts', (c) => {
    const result = services.contacts.listByResume(c.req.param('id'))
    if (!result.ok)
      return c.json({ error: result.error }, mapStatusCode(result.error.code))
    return c.json({ data: result.data })
  })

  return app
}

File: crates/forge-server/src/routes/resumes.rs
//! Resume CRUD routes with entry, section, and reorder sub-resources.
//!
//! Mirrors `packages/core/src/routes/resumes.ts`.

use axum::extract::{Path, Query, State};
use axum::routing::{get, post};
use axum::{Json, Router};
use serde::Deserialize;

use forge_core::{
    AddResumeEntry, CreateResume, Resume, ResumeEntry, ResumeSectionEntity, ResumeWithEntries,
    UpdateResume,
};
use forge_sdk::db::ResumeStore;

use crate::db::with_conn;
use crate::error::ApiError;
use crate::response::{ApiData, ApiList, Created, NoContent};
use crate::state::SharedState;

// ── Query params ────────────────────────────────────────────────────

#[derive(Debug, Deserialize, Default)]
pub struct ResumeListQuery {
    pub offset: Option<i64>,
    pub limit: Option<i64>,
}

// ── Request bodies ──────────────────────────────────────────────────

#[derive(Debug, Deserialize)]
pub struct UpdateEntryBody {
    pub content: Option<Option<String>>,
    pub section_id: Option<String>,
    pub position: Option<i32>,
}

#[derive(Debug, Deserialize)]
pub struct ReorderEntriesBody {
    pub entries: Vec<ReorderEntryItem>,
}

#[derive(Debug, Deserialize)]
pub struct ReorderEntryItem {
    pub entry_id: String,
    pub section_id: String,
    pub position: i32,
}

#[derive(Debug, Deserialize)]
pub struct CreateSectionBody {
    pub title: String,
    pub entry_type: String,
    pub position: Option<i32>,
}

// ── Handlers ────────────────────────────────────────────────────────

async fn create_resume(
    State(state): State<SharedState>,
    Json(input): Json<CreateResume>,
) -> Result<Created<Resume>, ApiError> {
    let result = with_conn(&state, move |conn| ResumeStore::create(conn, &input)).await?;
    Ok(Created(result))
}

async fn list_resumes(
    State(state): State<SharedState>,
    Query(q): Query<ResumeListQuery>,
) -> Result<Json<ApiList<Resume>>, ApiError> {
    let offset = q.offset.unwrap_or(0).max(0);
    let limit = q.limit.unwrap_or(50).clamp(1, 200);

    let (data, pagination) =
        with_conn(&state, move |conn| ResumeStore::list(conn, offset, limit)).await?;

    Ok(Json(ApiList { data, pagination }))
}

async fn get_resume(
    State(state): State<SharedState>,
    Path(id): Path<String>,
) -> Result<Json<ApiData<ResumeWithEntries>>, ApiError> {
    let result = with_conn(&state, move |conn| {
        ResumeStore::get_with_entries(conn, &id)?.ok_or_else(|| {
            forge_core::ForgeError::NotFound {
                entity_type: "Resume".into(),
                id: id.clone(),
            }
        })
    })
    .await?;
    Ok(Json(ApiData { data: result }))
}

async fn update_resume(
    State(state): State<SharedState>,
    Path(id): Path<String>,
    Json(input): Json<UpdateResume>,
) -> Result<Json<ApiData<Resume>>, ApiError> {
    let result =
        with_conn(&state, move |conn| ResumeStore::update(conn, &id, &input)).await?;
    Ok(Json(ApiData { data: result }))
}

async fn delete_resume(
    State(state): State<SharedState>,
    Path(id): Path<String>,
) -> Result<NoContent, ApiError> {
    with_conn(&state, move |conn| ResumeStore::delete(conn, &id)).await?;
    Ok(NoContent)
}

// ── Entry sub-resource handlers ────────────────────────────────────

async fn add_entry(
    State(state): State<SharedState>,
    Path(resume_id): Path<String>,
    Json(input): Json<AddResumeEntry>,
) -> Result<Created<ResumeEntry>, ApiError> {
    let result =
        with_conn(&state, move |conn| ResumeStore::add_entry(conn, &resume_id, &input))
            .await?;
    Ok(Created(result))
}

async fn update_entry(
    State(state): State<SharedState>,
    Path((resume_id, entry_id)): Path<(String, String)>,
    Json(body): Json<UpdateEntryBody>,
) -> Result<Json<ApiData<ResumeEntry>>, ApiError> {
    let result = with_conn(&state, move |conn| {
        ResumeStore::update_entry(
            conn,
            &resume_id,
            &entry_id,
            body.content,
            body.section_id,
            body.position,
        )
    })
    .await?;
    Ok(Json(ApiData { data: result }))
}

async fn remove_entry(
    State(state): State<SharedState>,
    Path((resume_id, entry_id)): Path<(String, String)>,
) -> Result<NoContent, ApiError> {
    with_conn(&state, move |conn| {
        ResumeStore::remove_entry(conn, &resume_id, &entry_id)
    })
    .await?;
    Ok(NoContent)
}

async fn reorder_entries(
    State(state): State<SharedState>,
    Path(resume_id): Path<String>,
    Json(body): Json<ReorderEntriesBody>,
) -> Result<NoContent, ApiError> {
    let entries: Vec<(String, String, i32)> = body
        .entries
        .into_iter()
        .map(|e| (e.entry_id, e.section_id, e.position))
        .collect();

    with_conn(&state, move |conn| {
        ResumeStore::reorder_entries(conn, &resume_id, &entries)
    })
    .await?;
    Ok(NoContent)
}

// ── Section sub-resource handlers ──────────────────────────────────

async fn create_section(
    State(state): State<SharedState>,
    Path(resume_id): Path<String>,
    Json(body): Json<CreateSectionBody>,
) -> Result<Created<ResumeSectionEntity>, ApiError> {
    let result = with_conn(&state, move |conn| {
        ResumeStore::create_section(conn, &resume_id, &body.title, &body.entry_type, body.position)
    })
    .await?;
    Ok(Created(result))
}

async fn delete_section(
    State(state): State<SharedState>,
    Path((resume_id, section_id)): Path<(String, String)>,
) -> Result<NoContent, ApiError> {
    with_conn(&state, move |conn| {
        ResumeStore::delete_section(conn, &resume_id, &section_id)
    })
    .await?;
    Ok(NoContent)
}

// ── Router ──────────────────────────────────────────────────────────

pub fn router() -> Router<SharedState> {
    Router::new()
        .route("/resumes", post(create_resume).get(list_resumes))
        .route(
            "/resumes/{id}",
            get(get_resume).patch(update_resume).delete(delete_resume),
        )
        .route("/resumes/{id}/entries", post(add_entry))
        .route(
            "/resumes/{resume_id}/entries/{entry_id}",
            axum::routing::patch(update_entry).delete(remove_entry),
        )
        .route("/resumes/{id}/entries/reorder", post(reorder_entries))
        .route("/resumes/{id}/sections", post(create_section))
        .route(
            "/resumes/{resume_id}/sections/{section_id}",
            axum::routing::delete(delete_section),
        )
}
